### PR TITLE
Feature/new force gun

### PR DESCRIPTION
**Pull request by JorisWillig**

### CODE REVIEW

File: game/systems/WeaponSystem.cpp [hunk 1]
```diff
@@ -21,11 +21,51 @@
...
+    projectileBody->setDensity(2);
+    projectileBody->setGravityScale(2);
+    projectileBody->setBullet(true);
+    ecsWorld->addComponent<BodyComponent>(projectileEntity, BodyComponent(std::move(projectileBody)));
+
+    auto dimensionComponent = DimensionComponent(dimensionVector);
+    ecsWorld->addComponent<DimensionComponent>(projectileEntity, dimensionComponent);
+
+    auto projectileComponent = ProjectileComponent(game::definitions::WeaponType::GrenadeLauncher, game::definitions::ProjectileType::Grenade);
+    //projectileComponent.force = direction;
```
Comment (RvdM7) [inline game/systems/WeaponSystem.cpp]: remove commented out code

File: engine/ecs/World.cpp [hunk 2]
```diff
@@ -13,7 +13,7 @@ namespace ecs {
         try {
             return m_entityManager.entity(id);
         } catch (...) {
-            throw; // probably entity not found
+            throw; // probably container not found
```
Comment (RvdM7) [inline engine/ecs/World.cpp]: After looking through your PR i think you may have mass renamed "entity" to "container", like here

File: engine/ecs/World.cpp [hunk 3]
```diff
@@ -37,7 +37,7 @@ namespace ecs {
                 auto& entity = getEntity(id);
                 destroyEntity(entity);
             } catch (...) {
-                // entity not found
+                // container not found
```
Comment (RvdM7) [inline engine/ecs/World.cpp]: Probably also mass renamed

File: game/builders/CharacterBuilder.cpp [hunk 4]
```diff
@@ -97,15 +97,15 @@ namespace builders {
             };
             m_ecsWorld.addComponent<components::BodyComponent>(players[i], std::move(bodyComponent));
 
-            // Create the dimension component for player entity
+            // Create the dimension component for player container
```
Comment (RvdM7) [inline game/builders/CharacterBuilder.cpp]: Probably also mass renamed

File: game/builders/CharacterBuilder.cpp [hunk 5]
```diff
@@ -97,15 +97,15 @@ namespace builders {
...
             components::DimensionComponent dimensionComponent{ common::Vector2D<double>(1.4, 2.1) };
             m_ecsWorld.addComponent<components::DimensionComponent>(players[i], dimensionComponent);
 
-            // Create the position component for player entity
+            // Create the position component for player container
```
Comment (RvdM7) [inline game/builders/CharacterBuilder.cpp]: Probably also mass renamed

File: game/builders/CharacterBuilder.cpp [hunk 6]
```diff
@@ -97,15 +97,15 @@ namespace builders {
...
             components::PositionComponent positionComponent{ position };
             m_ecsWorld.addComponent<components::PositionComponent>(players[i], positionComponent);
 
-            // Create the sprite component for the player entity
+            // Create the sprite component for the player container
```
Comment (RvdM7) [inline game/builders/CharacterBuilder.cpp]: Probably also mass renamed

File: game/builders/CharacterBuilder.cpp [hunk 7]
```diff
@@ -116,7 +116,7 @@ namespace builders {
             components::PlayerNameComponent name{ "P" + std::to_string((int)i + 1) };
             m_ecsWorld.addComponent<components::PlayerNameComponent>(players[i], name);
 
-            // Create the animations storage for the player entity
+            // Create the animations storage for the player container
```
Comment (RvdM7) [inline game/builders/CharacterBuilder.cpp]: Probably also mass renamed

File: game/builders/CharacterBuilder.cpp [hunk 8]
```diff
@@ -133,7 +133,7 @@ namespace builders {
                 m_ecsWorld.addComponent<components::PlayerInputComponent>(players[i], playerInputComponent);
             }
 
-            // Create the life component for player entity
+            // Create the life component for player container
```
Comment (RvdM7) [inline game/builders/CharacterBuilder.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 9]
```diff
@@ -46,11 +46,11 @@ namespace level {
 
             auto& entity = world.createEntity();
 
-            // Add a position component to character spawn entity
+            // Add a position component to character spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 10]
```diff
@@ -46,11 +46,11 @@ namespace level {
...
             auto posComponent = components::PositionComponent(position);
             world.addComponent<components::PositionComponent>(entity, posComponent);
 
-            // Add a character spawn component to character spawn entity
+            // Add a character spawn component to character spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 11]
```diff
@@ -59,21 +59,21 @@ namespace level {
...
-            // Add a sprite component to equipment spawn entity
+            // Add a sprite component to equipment spawn container
             auto spriteComponentPair = miscSpriteComponentMap.find("EquipmentSpawner");
             if (spriteComponentPair != miscSpriteComponentMap.end()) {
                 auto spriteComponent = spriteComponentPair->second;
                 world.addComponent<components::SpriteComponent>(entity, spriteComponent);
             }
 
-            // Add an equipment spawner component to equipment spawn entity
+            // Add an equipment spawner component to equipment spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 12]
```diff
@@ -59,21 +59,21 @@ namespace level {
         for (const auto& curSpawn : level.EquipmentSpawns) {
             common::Vector2D<double> position{ curSpawn.x, curSpawn.y };
             auto& entity = world.createEntity();
-            // Add a position component to equipment spawn entity
+            // Add a position component to equipment spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 13]
```diff
@@ -59,21 +59,21 @@ namespace level {
...
             auto posComponent = components::PositionComponent(position);
             world.addComponent<components::PositionComponent>(entity, posComponent);
-            // Add a dimension component to equipment spawn entity
+            // Add a dimension component to equipment spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 14]
```diff
@@ -59,21 +59,21 @@ namespace level {
...
             auto dimensionComponent = components::DimensionComponent(common::Vector2D<double>(1.2, 0.4));
             world.addComponent<components::DimensionComponent>(entity, dimensionComponent);
 
-            // Add a sprite component to equipment spawn entity
+            // Add a sprite component to equipment spawn container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 15]
```diff
@@ -108,11 +108,11 @@ namespace level {
                     world.addComponent<components::TileComponent>(entity);
                     common::Vector2D<double> position{ nextTile->x, nextTile->y };
 
-                    // Add a position component to tile entity
+                    // Add a position component to tile container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 16]
```diff
@@ -108,11 +108,11 @@ namespace level {
...
                     auto posComponent = components::PositionComponent(position);
                     world.addComponent<components::PositionComponent>(entity, posComponent);
 
-                    // Add a sprite component to tile entity
+                    // Add a sprite component to tile container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 17]
```diff
@@ -131,16 +131,16 @@ namespace level {
             auto& entity = world.createEntity();
             common::Vector2D<double> position{ currentTile.x, currentTile.y };
 
-            // Add a position component to tile entity
+            // Add a position component to tile container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 18]
```diff
@@ -131,16 +131,16 @@ namespace level {
...
             auto posComponent = components::PositionComponent(position);
             world.addComponent<components::PositionComponent>(entity, posComponent);
             world.addComponent<components::TileComponent>(entity);
 
-            // Add a body component to tile entity
+            // Add a body component to tile container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/level/reader/LevelReader.cpp [hunk 19]
```diff
@@ -131,16 +131,16 @@ namespace level {
...
             auto bodyComponent = components::BodyComponent(physics.createStaticBody(position, bodySize, entity.id()));
             world.addComponent<components::BodyComponent>(entity, std::move(bodyComponent));
 
-            // Add a sprite component to tile entity
+            // Add a sprite component to tile container
```
Comment (RvdM7) [inline game/level/reader/LevelReader.cpp]: Probably also mass renamed

File: game/systems/CheatsSystem.cpp [hunk 20]
```diff
@@ -11,7 +11,7 @@ bool executeInfiniteJump(engine::ecs::World* ecsWorld,
     engine::ecs::Entity& entity)
 {
     if (entity.hasComponent<game::components::JumpComponent>()) {
-        // auto& jumpComponent = ecsWorld->getComponent<game::components::JumpComponent>(entity);
+        // auto& jumpComponent = ecsWorld->getComponent<game::components::JumpComponent>(container);
```
Comment (RvdM7) [inline game/systems/CheatsSystem.cpp]: Probably also mass renamed

File: game/systems/InventorySystem.cpp [hunk 21]
```diff
@@ -41,7 +41,7 @@ namespace systems {
     {
         double equipableCandidateDistance = -1;
         engine::ecs::Entity* equipableCandidate = nullptr;
-        // find the closest entity near the player that can be equipped
+        // find the closest container near the player that can be equipped
```
Comment (RvdM7) [inline game/systems/InventorySystem.cpp]: Probably also mass renamed

File: game/systems/items/ReverseGravitySystem.cpp [hunk 22]
```diff
@@ -35,7 +35,7 @@ namespace systems {
                                                        .count()
                             / 1000.0;
                         if (secondsSinceFired > DURATION_SECONDS) {
-                            // disable effect and remove item entity
+                            // disable effect and remove item container
```
Comment (RvdM7) [inline game/systems/items/ReverseGravitySystem.cpp]: Probably also mass renamed

File: tests/engine/ecs/ecs_test.cpp [hunk 23]
```diff
@@ -55,16 +55,16 @@ TEST_CASE("Entities can have a component attached to them", "[ecs]")
     ecs::World world;
     auto& entity = world.createEntity();
 
-    SECTION("The entity does not have the component yet")
+    SECTION("The container does not have the component yet")
```
Comment (RvdM7) [inline tests/engine/ecs/ecs_test.cpp]: Probably also mass renamed

File: tests/engine/ecs/ecs_test.cpp [hunk 24]
```diff
@@ -55,16 +55,16 @@ TEST_CASE("Entities can have a component attached to them", "[ecs]")
...
     {
         REQUIRE_FALSE(entity.hasComponent<TestComponent>());
     }
-    SECTION("When adding a component, the entity will register it")
+    SECTION("When adding a component, the container will register it")
```
Comment (RvdM7) [inline tests/engine/ecs/ecs_test.cpp]: Probably also mass renamed

File: tests/engine/ecs/ecs_test.cpp [hunk 25]
```diff
@@ -55,16 +55,16 @@ TEST_CASE("Entities can have a component attached to them", "[ecs]")
...
     {
         world.addComponent<TestComponent>(entity, TestComponent{});
         REQUIRE(entity.hasComponent<TestComponent>());
     }
-    SECTION("When adding multiple different components, the entity will register it")
+    SECTION("When adding multiple different components, the container will register it")
```
Comment (RvdM7) [inline tests/engine/ecs/ecs_test.cpp]: Probably also mass renamed

File: tests/engine/ecs/ecs_test.cpp [hunk 26]
```diff
@@ -95,7 +95,7 @@ TEST_CASE("ECS World can iterate over entities", "[ecs]")
         }
 
         std::size_t iteratedEntities = 0;
-        world.forEachEntityWith<TestComponent>([&](const auto& /*entity*/) {
+        world.forEachEntityWith<TestComponent>([&](const auto& /*container*/) {
```
Comment (RvdM7) [inline tests/engine/ecs/ecs_test.cpp]: Probably also mass renamed

File: game/listeners/ProjectileContactListener.h [hunk 27]
```diff
@@ -5,34 +5,43 @@
...
+        void applyForce(engine::ecs::Entity& body, DataContainer projectile);
+        void explode(DataContainer projectile, double radius);
         engine::ecs::Entity* findPortal(bool alternative);
 
     public:
         explicit ProjectileContactListener(engine::ecs::World* world, engine::physics::World* physics)
-            : m_ecsWorld{ world }
-            , m_physicsWorld{ physics } {};
+            : m_ecsWorld { world }
+            , m_physicsWorld { physics } {};
```
Comment (RvdM7) [inline game/listeners/ProjectileContactListener.h]: According to my clang, there should not be spaces
nitpicky but still

File: game/systems/AnimationSystem.cpp [hunk 28]
```diff
@@ -42,7 +42,8 @@ namespace systems {
                 if (elapsedSeconds >= spriteComponent.frameTime) {
                     advanceFrame(spriteComponent);
                 }
-                m_sprites.push_back(sprite);
+                m_sprites.emplace_back(sprite);
+                //std::sort(m_sprites.begin(), m_sprites.end(), compareFunc);
```
Comment (RvdM7) [inline game/systems/AnimationSystem.cpp]: please remove commented out code, or add it in for extra lagg
but seriously, remove, ill add it back in when its fixed

File: game/systems/WeaponSystem.cpp [hunk 29]
```diff
@@ -21,11 +21,51 @@
...
+    ecsWorld->addComponent<ProjectileComponent>(projectileEntity, projectileComponent);
+    ecsWorld->addComponent<OnOutOfBoundsDeleteComponent>(projectileEntity);
+    ecsWorld->getComponent<BodyComponent>(entity).body->applyForce(common::Vector2D<double>((direction.x > 0 ? -600 : 600), 0), playerPosition);
```
Comment (RvdM7) [inline game/systems/WeaponSystem.cpp]: after testing for some bit, i feel like the projectile knockback range is larger than the actual seen effect
this line of code can be commented out to have the player only move when moved by input, or when hit with a weapon, just as a little helper when testing

Comment (RvdM7) [reply game/systems/WeaponSystem.cpp]: After further testing this does not seem to be the case
but the effect makes it look like it is
you could change the effect, or not

File: engine/ecs/World.cpp (hunk 2 above)
Comment (JorisWillig) [reply engine/ecs/World.cpp]: Woopsiedaisies....

**Pull request merged**